core/dataset.cpp

```diff
@@ -773,24 +780,24 @@ Dataset operator+(const Dataset &lhs, const DataConstProxy &rhs) {
   return res;
 }
 
-Dataset operator+(Dataset &&lhs, const Dataset &rhs) {
-  return apply(plus_equals, lhs, rhs);
+Dataset &&operator+(Dataset &&lhs, const Dataset &rhs) {
```

Comment:
Return by value!

Reply:
By doing so you force a copy, is avoiding this not the point of supporting rvalue arguments?

Reply:
Arguments yes, return values no, see RVO (return value optimization):

https://stackoverflow.com/questions/12953127/what-are-copy-elision-and-return-value-optimization

Also, even if RVO did not happen, it would be a move, not a copy, i.e., cheap.

Furthermore, if you return an rvalue reference you might get a dangling reference since the object it points to might get destroyed.